### PR TITLE
[WIP] Replace global data.json with page-data.json per page

### LINKED ISSUE #4746: Split page metadata so can lazy load it and reduce the initial JS
Summary: Allow the `data.json` file to be loaded in shards for sites with large numbers of pages.

## History

Once #4715 lands:

* `data.json` contains a map of all `path`s to data file names
  - File names include a content hash
* `data.json` is loaded after the page has finished rendering

Sharding was also discussed in #4626.

## Changes

Rather than loading `data.json` as a single file, it could be split into multiple files based on a simple sharding mechanism, for example:

```es6
const calculateShardNameForId(shardLength: int, id: string) => md5(id).substr(0, shardLength)
```

@KyleAMathews [mentioned](https://github.com/gatsbyjs/gatsby/issues/4626#issuecomment-375439870):

>  Ideally you'd shard by something like path names so a shard for /blog/*, as those are likely to be needed together.

@pieh [shared](https://github.com/gatsbyjs/gatsby/issues/4626#issuecomment-376946243):

> Just to give more info - when I run my tests against https://github.com/freeCodeCamp/guides (~2800 pages) - gzipped "webpackified" data.json is 141KB - this surely won't scale very nice for 100 000 pages sites, but up to 10 000 I think this shouldn't be that much of a issue

## Rationale

* On very large sites (>1k pages, or >10k pages) the `data.json` file can become large.
* It must be fully loaded before prefetching can begin.
* Splitting it into chunks could allow faster prefetching, and require less data over the wire, on very large sites.

This is a low priority enhancement.

### PR DESCRIPTION
**Pull request by Moocar**
## Description

Removes the global `data.json` file and instead provides a `page-data.json` for each page on a gatsby site. See https://github.com/gatsbyjs/gatsby/issues/11982 for background.

This PR is a draft and should not be merged. I'll be providing a series of smaller PRs instead. Reviewers can always reference this PR to see the full context of the changes. That said, if you're reading this PR and spot any problems, then please do leave some feedback!

## Benefits/Trade offs

**pros**

- no more downloading `data.json` to the browser. For 10'000 page sites, this file was already approaching 500kb compressed.
- navigating to prefetched pages should be faster since there's no `data.json` to search through
- `production-app.js` webpack build no longer references pages or their data. Which will make incremental builds much easier to implement in the future

**cons**

- Requesting a page from the server now requires 2 requests. First `page-data.json`, then the component. Though prefetched pages still do this in the background.
- Since page data isn't immutable anymore, we can't rely on infinite browser caching
- Some potential breaking changes in `loader.getResourceURLsForPathname`

## Sup-PRs

Breaking up these PRs is hard. To make it a bit easier, I'm going to have gatsby saving two sources of data to disk. `data.json` and `static/d/...`, as well as the new `page-data.json`. This will require duplication of memory and disk, so we won't be able to merge to master. But we will be able to gradually introduce new functionality, and once it's ready, delete the old data sources and merge to master. I'm not a fan of long lived non-master branches, but I think it makes sense here.

The long lived branch is [per-page-manifest](https://github.com/gatsbyjs/gatsby/tree/per-page-manifest). The PRs that will merge directly into that are:

**TODO**

- [ ] Bad handling of timeouts etc https://sentry.io/organizations/gatsby-inc-9z/issues/1054501319/?project=1471074&statsPeriod=24h
- [X] page-data.json requested a LOT (resulting in 304s)
  - From my reading, it looks like these are all coming from sentry's `breadcrumbs.js` file.

**TODO but won't block release**

- [ ] https://github.com/gatsbyjs/gatsby/issues/13430
- [ ] Review "behind the scenes" docs. They'll be very out of date now.
- [ ] `public/static/d/....` isn't used except for static query results. So we should ask the user to delete their `public` directory when they run gatsby with this code for the first time? It's not required.
- [ ] [Add prefetch for linked component modules to page HTML](https://github.com/gatsbyjs/gatsby/issues/14262)

Already reviewed and Merged:

- [x] [~~Merge build-html/develop-htm~~l](https://github.com/gatsbyjs/gatsby/pull/13005/files)
- [x] [~~gatsby develop: cleanup port detection~~](https://github.com/gatsbyjs/gatsby/pull/13006)
- [x] [~~expicitly call db saveState and autoSave~~](https://github.com/gatsbyjs/gatsby/pull/13007)
- [x] [~~move redirects-writer to bootstrap~~](https://github.com/gatsbyjs/gatsby/pull/13008)
- [x] [~~write match-paths.json~~](https://github.com/gatsbyjs/gatsby/pull/13012)
- [x] [~~move src/internal-plugins/query-runner to src/query~~](https://github.com/gatsbyjs/gatsby/pull/13016)
- [X] [~~explicitly watch delete page from gatsby develop~~](https://github.com/gatsbyjs/gatsby/pull/13099)
- [x] [~~Make query queue constructable~~](https://github.com/gatsbyjs/gatsby/pull/13061)
- [X] [~~Refactor query calc query ids~~](https://github.com/gatsbyjs/gatsby/pull/13063)
- [X] [~~Get list of /dev-404-page/ pages via graphql~~](https://github.com/gatsbyjs/gatsby/pull/13186)
- [X] [~~Split static/page queries~~](https://github.com/gatsbyjs/gatsby/pull/13064)
- [X] [~~Move query/page-query-runner to query/index~~](https://github.com/gatsbyjs/gatsby/pull/13375)

Merged against `per-page-manifest`
- [X] [~~Page data without compilation hash~~](https://github.com/gatsbyjs/gatsby/pull/13139) (ready for review)
- [X] [~~Websocket manager use page data~~](https://github.com/gatsbyjs/gatsby/pull/13389)
- [X] [~~loader-use-page-data~~](https://github.com/gatsbyjs/gatsby/pull/13409)
- [X] [~~remove-data-json-from-guess~~](https://github.com/Moocar/gatsby/compare/Moocar:loader-use-page-data...Moocar:remove-data-json-from-guess)
- [X] [~~gatsby-serve-use-page-data~~](https://github.com/Moocar/gatsby/compare/Moocar:loader-use-page-data...Moocar:gatsby-serve-use-page-data)
- [X] [~~remove-data-json~~](https://github.com/gatsbyjs/gatsby/pull/13729) 
- [X] [~~Refresh browser if webpack rebuild occurs~~](https://github.com/gatsbyjs/gatsby/pull/13871)
- [X] [~~Remove json-name~~](https://github.com/gatsbyjs/gatsby/pull/14167)
- [X] [~~loader.js getPage -> loadPageSync~~](https://github.com/gatsbyjs/gatsby/pull/14264)
- [X] [~~Page data loading resilience~~](https://github.com/gatsbyjs/gatsby/pull/14286)

## Related Issues

- Related to https://github.com/gatsbyjs/gatsby/pull/12711
- Addresses https://github.com/gatsbyjs/gatsby/issues/11982
- https://github.com/gatsbyjs/gatsby/issues/13430
- Fixes #4746

### CODE REVIEW

**Comment by Moocar** [comment]
@bennetthardwick If you have some time, I'd very much appreciate any testing you can provide on this one. It should result in a faster initial load, and faster page navigation (assuming prefetching has occurred). 

Bare in mind though, it's super new code and touches a lot of Gatsby, so I'm anticipating bugs.

File: packages/gatsby/src/redux/reducers/match-paths.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  switch (action.type) {
+    case `DELETE_CACHE`:
+      return {}
+    case `CREATE_PAGE`: {
+      const page = action.payload
+      if (page.matchPath) {
+        const { oldPage } = action
+        const newState = { ...state }
+        if (oldPage && oldPage.matchPath !== page.matchPath) {
+          delete newState.matchMaths[oldPage.matchPath]
```
Comment (bennetthardwick) [inline packages/gatsby/src/redux/reducers/match-paths.js]: This mutates the original `state` object, is that okay?

**Comment by bennetthardwick** [comment]
Hey @Moocar this is awesome! On 25k pages there are no visible performance issues. I'll keep testing throughout the week and let you know if I find anything.

File: packages/gatsby/src/redux/reducers/match-paths.js (hunk 1 above)
Comment (Moocar) [reply packages/gatsby/src/redux/reducers/match-paths.js]: Good catch. Not to mention that `matchMaths` makes no sense whatsoever 😂. Fixed in https://github.com/gatsbyjs/gatsby/pull/13012

**Comment by andrewfairlie** [comment]
I have tested this build locally on a site with just over 11,000 pages. It has made a **huge** difference and will help us scale Gatsby dramatically.

**Comment by Moocar** [comment]
Thanks so much for the testing @bennetthardwick and @andrewfairlie! Now to get it shipped

**Comment by KyleAMathews** [comment]
@andrewfairlie would love to hear more about what changed!

**Comment by andrewfairlie** [comment]
Of course. 

On a single entry change in our CMS on master we had 11,418 files change after build which would need to upload to Netlify.

With this new method one entry change in the CMS had only 2 file changes. One for the HTML and one for the page-data JSON. 

We work on some large publishing sites, so only having to upload these tiny changes on build would help us get content to the reader sooner and avoid Netlify timing out which is currently an issue.

**Comment by polarathene** [comment]
> With this new method one entry change in the CMS had only 2 file changes. One for the HTML and one for the page-data JSON.

So this change will make it possible to only build pages where actual changes have occurred? 

This was one of my main concerns for using Gatsby on large community sites which cater to user submitted content(such as a profile or similar template with user submitted data).  It did not seem practical to rebuild every single page when a user modified their content or submitted new content. With this new approach, is using Gatsby for community driven sites like I described more viable?

**Comment by andrewfairlie** [comment]
@polarathene I don't think it stops every page from building. But it should mean that the built pages are identical unless the content and/or code has changed and so won't need to be uploaded again (which is a big win on services like Netlify where upload and post-processing can take a while)

For your needs, it might be worth considering not building community-driven pages and doing that on client-side. Gatsby is capable of being both. This blog post might be useful https://www.gatsbyjs.org/blog/2018-11-07-gatsby-for-apps/#gmail

**Comment by wardpeet** [comment]
@polarathene @andrewfairlie this actually gives us the opportunity to increments builds and create pages on the fly. In the current version, you'll need to rebuild all your pages because of data.json knows every page right now. With this PR we can get the assets per page so each page can fetch it's own dependencies which makes everything much more scalable.

@Moocar feel free to fix my comment and explain things better 😄

**Comment by andrewfairlie** [comment]
That's brilliant, I didn't appreciate that aspect of it.

**Comment by polarathene** [comment]
> For your needs, it might be worth considering not building community-driven pages and doing that on client-side.

@andrewfairlie If I understand you correctly, you mean to request data from the server after initial page load on the client to grab any "dynamic" specific content for that page (text/images), using it as a static template(AppShell) for the rest?

I suppose that'd work, but I'd miss out on some build-time features/benefits of building static content to serve? Especially with images, `gatsby-image` would require having all it's metadata stored somewhere and fetched(which I get happens implicitly with built pages), There would be no base64 placeholder inlined, and I'd need extra handling for waiting on the data response, else have content jump around?

I've been considering to assist with a friends community site moving from PHP and it's forum software roots to a modern web stack which'd be easier to work with and perform better. [Here's an example of user content](https://modworkshop.net/mydownloads.php?action=view_down&did=24675), some of the data suits being fetched after load on the client-side(Stats, comments), but there is quite a bit that'd be infrequently updated and would benefit from Gatsby.

I've been under the impression that if a user submits or updates content like what I had linked to, Gatsby would currently do a full rebuild of all pages, or similar concerns vs building such a site with say Next.js. 

I don't think I've seen Gatsby promoted for a scenario where it could build/update pages to serve on demand(Gatsby Preview seems to do this but for different reasons). I was curious if this PR actually makes this usecase viable with Gatsby.

Thanks for the link though, I wasn't aware of being able to use Gatsby like that.

> this actually gives us the opportunity to increments builds and create pages on the fly.

@wardpeet So I could use Gatsby on a server to generate new/updated pages frequently from user content without having to render on-demand server-side(and then cache for future requests)?

I'll be assisting with a rewrite of a friends community website later this year and was under the impression that Gatsby wasn't suitable for such a site with user curated content pages(the pages themselves can be rather static, it's just that new content/pages would be added or existing ones infrequently updated). [Example user content page](https://modworkshop.net/mydownloads.php?action=view_down&did=24675)(ignore dynamic data like stats/comments)

Seemed a bit silly if each time a user changed/added content for a specific page it would require re-building the whole site, so I figured Next.js or others might be more appropriate. 

It's currently running off PHP with a DB(each page making requests, I don't think they've added caching), making heavy use of the resources (quad core, 32GB RAM, 10TB bandwidth a month), I think there is plenty of room for bringing the resource needs down :smile:

**Comment by wardpeet** [comment]
We have the ability to go create a [full client side website](https://github.com/gatsbyjs/gatsby/tree/master/examples/client-only-paths) with an app shell model. You can also install the [gatsby-offline-plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-offline/?=offline) which loads the gatsby app with an app shell next load.

If you go full client side, you might lose some SEO benefits if the page rapidly updates as google only crawls js based websites once a week.

Gatsby will be able to only create pages that are changed or added so not all pages need to be rebuild. It will still need to do some bootstrapping to build the page so it's probably still good to have an async job that builds the website instead of doing it on users request. If you go for gatsby you probably want to be able to update your comments realtime with graphql on the client to so it might be more work than nextjs but gatsby is cheaper on hardware.

**Comment by alirussell** [comment]
Hi @Moocar,

Thanks for this. We are building 70k+ pages, and just realised that the pages manifest file was over 24mb uncompressed, which is not ideal.  It seems to work well with our new site in testing so far!

Is there any ETA on when you think you will get all the smaller PRs merged?  This does seem to fix our issue, and we are happy to use this for building to start with (we are under a tight deadline to get the new site up), but don't want to be using something experimental for too long?

Thanks

**Comment by Moocar** [comment]
@alirussell Good to hear! Judging by the responses to this issue, the changes seem to be working great. So barring any major design flaws that are found during the review process, I'd say it's just a matter of time before this is merged. Sorry I can't give you much more than that.

**Comment by dja** [comment]
@Moocar We're super excited about this as well. What's the latest on it?

**Comment by leonfs** [comment]
We are super excited too. How long do you think it will take to make its way to master?

**Comment by kevee** [comment]
A quick piece of feedback - we've been running a [very large Gatsby site](https://csumb.edu) for a week now with this PR and it's been working great. 

One quick note, any 404 page is trying to load `page-data.json,` which won't work because that page doesn't exist. So visiting `https://mysite.com/not-there` will have the 404 page try to load `page-data/not-there/page-data.json`.

**Comment by Moocar** [comment]
Thanks for your patience everyone. We finally have a release candidate ready for this work. I really appreciate all the testing and feedback that everyone on this thread has given so far. It's so good to get early feedback.

The release can be found on the [gatsby@per-page-manifest](https://www.npmjs.com/package/gatsby/v/2.5.1-per-page-manifest.68) dist-tag. To test it out:

```
npm install --save gatsby@per-page-manifest
gatsby build
```

This update also touches `gatsby-plugin-offline` and `gatsby-plugin-guess-js` so you'll have to install the `per-page-manifest` version for those too if your site uses them.

Would very much appreciate another round of testing on this latest release!

**Comment by Moocar** [comment]
> One quick note, any 404 page is trying to load `page-data.json,` which won't work because that page doesn't exist. So visiting `https://mysite.com/not-there` will have the 404 page try to load `page-data/not-there/page-data.json`.

@kevee This is necessary now since the browser no longer has a global list of all pages on the site. So it has to make a request to the server for the `/page-data/{url-path}/page-data.json` in order to find out if the page exists or not. Whereas in the old approach, it would look up the list of pages in memory and simply display the 404 page if it wasn't found.

**Comment by me4502** [comment]
Is the tagged release built from the `per-page-manifest` branch? Or does it include https://github.com/gatsbyjs/gatsby/pull/14167 ?

**Comment by Moocar** [comment]
@me4502 It's actually built from https://github.com/gatsbyjs/gatsby/pull/14167. Guilty :)

**Comment by me4502** [comment]
image

This is with 60k pages using the test build, definitely makes a massive improvement to the Lighthouse Performance section :)

**Comment by Moocar** [comment]
@me4502 woo! Out of interest, how does that compare to before?

**Comment by me4502** [comment]
It was on 96 for performance, but the main win is in the Performance metrics

image

These are now all ticks, and the diagnostics section is empty.

**Comment by KyleAMathews** [comment]
Nice! 📣

Did it make a difference on build times at all?

**Comment by me4502** [comment]
Actually yeah, 15 minutes down to 9. And I except it to have massively lowered deploy times due to less changed files. I haven't ran multiple deploys yet to test though.

**Comment by KyleAMathews** [comment]
@me4502 oh wow! That's kinda surprising actually — where'd the 6 minutes get saved?

**Comment by Moocar** [comment]
OK, we're ready to finally merge `per-page-manifest` into master. I just have one concern before we do. I've changed an official browser API [onPostPrefetchPathname -> onPostPrefetch](https://github.com/gatsbyjs/gatsby/pull/14359/files#diff-328e1665577217c4f69c4fd1b2f76258L205). It's only used by the `gatsby-plugin-offline` plugin (that we know of). So if we merge `per-page-manifest` to master, publish, and then a site upgrades `gatsby` but not `gatsby-plugin-offline`, they'll. get an error:

```
Your plugins must export known APIs from their gatsby-browser.js.
The following exports aren't APIs. Perhaps you made a typo or your plugin is outdated?

See https://www.gatsbyjs.org/docs/browser-apis/ for the list of Gatsby Browser APIs

- The plugin \"gatsby-plugin-offline@2.1.1\" is exporting a variable named \"onPostPrefetchPathname\" which isn't an API.

Perhaps you meant to export \"onPrefetchPathname\"?
error Command failed with exit code 1
```

Basically it's a breaking change. And we need users to update both `gatsby` and `gatsby-plugin-offline` at the same time. Which to my understanding isn't possible (but happy to be shown otherwise). I'll look at how much work it will be to make it backwards compatible.

**Comment by me4502** [comment]
@KyleAMathews Seems to just be generally all over the build:
Old:
```
[2019-05-20T03:34:23.683Z] success open and validate gatsby-configs — 0.015 s
[2019-05-20T03:34:23.944Z] success load plugins — 4.015 s
[2019-05-20T03:34:23.944Z] success onPreInit — 0.006 s
[2019-05-20T03:34:23.944Z] success delete html and css files from previous builds — 0.007 s
[2019-05-20T03:34:23.944Z] success initialize cache — 0.009 s
[2019-05-20T03:34:23.944Z] success start nodes db — 0.005 s
[2019-05-20T03:34:24.203Z] success copy gatsby files — 0.042 s
[2019-05-20T03:34:24.203Z] success onPreBootstrap — 0.014 s
[2019-05-20T03:34:24.203Z] Starting to fetch data from Prismic
[2019-05-20T03:34:24.462Z] Fetch Prismic data: 387.842ms
[2019-05-20T03:35:03.976Z] success source and transform nodes — 35.679 s
[2019-05-20T03:35:03.976Z] success building schema — 1.061 s
[2019-05-20T03:38:25.535Z] success createPages — 193.127 s
[2019-05-20T03:38:25.536Z] success createPagesStatefully — 0.075 s
[2019-05-20T03:38:25.536Z] success onPreExtractQueries — 0.027 s
[2019-05-20T03:38:25.536Z] success update schema — 0.000 s
[2019-05-20T03:38:25.536Z] success extract queries from components — 0.389 s
[2019-05-20T03:38:25.536Z] success run static queries — 0.099 s — 3/3 30.93 queries/second
[2019-05-20T03:39:31.986Z] success run page queries — 75.684 s — 55490/55490 736.00 queries/second
[2019-05-20T03:39:33.397Z] success write out page data — 2.811 s
[2019-05-20T03:39:33.397Z] success write out redirect data — 0.001 s
[2019-05-20T03:42:10.295Z] 
[2019-05-20T03:42:10.295Z] success Build manifest and related icons — 0.257 s
[2019-05-20T03:42:10.295Z] success onPostBootstrap — 0.260 s
[2019-05-20T03:42:10.295Z] 
[2019-05-20T03:42:10.295Z] info bootstrap finished - 467.029 s
[2019-05-20T03:42:10.295Z] 
[2019-05-20T03:43:47.136Z] success Building production JavaScript and CSS bundles — 76.443 s
[2019-05-20T03:47:23.755Z] success Building static HTML for pages — 210.809 s — 55490/55490 276.87 pages/second
[2019-05-20T03:47:23.755Z] Generated public/sw.js, which will precache 12 files, totaling 402207 bytes.
[2019-05-20T03:47:23.755Z] info Done building in 782.100299522 sec
[2019-05-20T03:47:23.755Z] Done in 841.10s.
[2019-05-20T03:47:23.755Z] Done in 841.64s.
```

New:
```
[2019-05-24T04:15:58.931Z] success open and validate gatsby-configs — 0.010 s
[2019-05-24T04:16:03.113Z] success load plugins — 4.288 s
[2019-05-24T04:16:03.113Z] success onPreInit — 0.005 s
[2019-05-24T04:16:03.113Z] success delete html and css files from previous builds — 0.006 s
[2019-05-24T04:16:03.113Z] success initialize cache — 0.008 s
[2019-05-24T04:16:03.113Z] success start nodes db — 0.004 s
[2019-05-24T04:16:03.113Z] success copy gatsby files — 0.033 s
[2019-05-24T04:16:03.113Z] success onPreBootstrap — 0.014 s
[2019-05-24T04:16:03.113Z] Starting to fetch data from Prismic
[2019-05-24T04:16:03.371Z] Fetch Prismic data: 391.532ms
[2019-05-24T04:16:36.213Z] success source and transform nodes — 29.421 s
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] warning Deprecation warning - "noDefaultResolvers" is deprecated. In Gatsby 3, defined fields won't get resolvers, unless explicitly added with a directive/extension.
[2019-05-24T04:16:36.213Z] success building schema — 0.507 s
[2019-05-24T04:19:27.717Z] success createPages — 163.235 s
[2019-05-24T04:19:27.717Z] success createPagesStatefully — 0.071 s
[2019-05-24T04:19:27.717Z] success onPreExtractQueries — 0.004 s
[2019-05-24T04:19:27.717Z] success update schema — 0.000 s
[2019-05-24T04:19:27.717Z] success extract queries from components — 0.376 s
[2019-05-24T04:19:27.717Z] success write out requires — 0.291 s
[2019-05-24T04:19:27.717Z] success write out redirect data — 0.001 s
[2019-05-24T04:19:27.717Z] success Build manifest and related icons — 0.117 s
[2019-05-24T04:19:27.717Z] success onPostBootstrap — 0.119 s
[2019-05-24T04:19:27.717Z] info bootstrap finished - 200.574 s
[2019-05-24T04:19:27.717Z] success run static queries — 0.011 s — 2/2 194.43 queries/second
[2019-05-24T04:19:35.850Z] success Building production JavaScript and CSS bundles — 17.572 s
[2019-05-24T04:19:36.108Z] success Rewriting compilation hashes — 0.066 s
[2019-05-24T04:21:42.166Z] success run page queries — 122.355 s — 55588/55588 454.93 queries/second
[2019-05-24T04:24:40.233Z] success Building static HTML for pages — 157.017 s — 55588/55588 381.18 pages/second
[2019-05-24T04:24:40.233Z] Generated public/sw.js, which will precache 12 files, totaling 404337 bytes.
[2019-05-24T04:24:40.233Z] info Done building in 515.287176085 sec
[2019-05-24T04:24:40.233Z] Done in 578.07s.
[2019-05-24T04:24:40.233Z] Done in 578.55s.
```

**Comment by KyleAMathews** [comment]
Building JavaScript is faster as expected — interesting that running queries & SSR is so much faster 🤔

--
Kyle Mathews

Blog: http://bricolage.io
Twitter: http://twitter.com/kylemathews

**Comment by Moocar** [comment]
OK, I'm satisfied again that this is ready to merge to master. I've published a new [gatsby@per-page-manifest](https://www.npmjs.com/package/gatsby/v/2.7.4-per-page-manifest.15).

I've created the PR to merge `per-page-manifest` into master. https://github.com/gatsbyjs/gatsby/pull/14359

I'm also running https://github.com/pieh/develop-runner to make sure that a bunch of sites at least build.

**Comment by Moocar** [comment]
Hey folks, I'm closing this PR, as the actual code to be merged to master is on https://github.com/gatsbyjs/gatsby/pull/14359 (branch `per-page-manifest`). Please comment there from now on.

**Comment by KyleAMathews** [comment]
For those following along here, this PR got merged via https://github.com/gatsbyjs/gatsby/pull/14359 and released as 2.9.0 🎉 Thanks @Moocar for the great analysis and code and @wardpeet for all your reviews!

**Comment by leonfs** [comment]
Thank-you everyone!! amazing job!!
